File: 33th/jongmlee/Truck.java
import java.util.*;

class Solution {
	public int solution(int bridge_length, int weight, int[] truck_weights) {
		int time = 0;
		int sum = 0;
		Queue<Integer> q = new LinkedList<>();
		for (int tw : truck_weights) {
			while (true) {
				if (q.isEmpty()) {
					q.add(tw);
					sum += tw;
					time++;
					break;
				} else if (q.size() == bridge_length) {
					int truck = q.poll();
					sum -= truck;
					time++;
					if (tw + sum <= weight) {
						q.add(tw);
						sum += tw;
						break;
					} else {
						q.add(0);
					}
				} else {
					time++;
					if (tw + sum <= weight) {
						q.add(tw);
						sum += tw;
						break;
					} else {
						q.add(0);
					}
				}
			}
		}
		return time + bridge_length;
	}
}

File: 33th/jeakim/truck.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(int bridge_length, int weight, vector<int> truck_weights) {
    int answer = 0;
	int cw = 0;
	queue<int> q;
	int time[10000] = {0};
	int	i = 0;

	q.push(i);
	while (int i = 1; i < truck_weights.size(); i++){
		if (cw < weight && q.size() < bridge_length){
			q.push(i);
			time[i]++;
		}
		if (time[q.front()] > bridge_length)
			q.pop();
		answer++;
	}
    return answer;
}